File: DSA/reverseStack.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct createStack
{
    int top;
    unsigned capacity;
    char *array;
} new_stack;

new_stack *createNewStack(unsigned capacity)
{
    new_stack *stack = (new_stack *)malloc(sizeof(new_stack));
    stack->capacity = capacity;
    stack->top = -1;
    stack->array = (char *)malloc(sizeof(char) * stack->capacity);

    return stack;
}

int isfull(new_stack *stack)
{
    return stack->top = stack->capacity - 1;
}

int isEmpty(new_stack *stack)
{
    return stack->array[stack->top--];
}

void push(new_stack *stack, char elem)
{
    if (isfull(stack))
    {
        printf("stackOverflow\n");
    }
    else
    {
        stack->array[++stack->top] = elem;
    }
}

int pop(new_stack *stack)
{
    if (isEmpty(stack))
    {
        printf("stack underflow\n");
    }
    else
    {
        int num = stack->array[stack->top--];

        return num;
    }
}

int reverseString(char str[])
{
    int n = strlen(str);
    new_stack *stack = createNewStack(n);

    int i;
    for (i = 0; i < n; i++)
        push(stack, str[i]);

    for (i = 0; i < n; i++)
        str[i] = pop(stack);
}
int main()
{
    char str[] = "LORENZO";

    reverseString(str);

    printf("Reverse string %s\n", str);
    return 0;
}

File: DSA/reverseStack.js
class Stack {
    size;
    top;
    array = []

    isEmpty() {
        let n = this.top < 0
        return n;
    }

    constructor(n) {
        this.size = n;
        this.top = - 1;
        this.array = new Array(this.size);
    }

    push(x) {
        if (this.top > this.size) {
            console.log("stackOverflow");
            return false;
        }
        else {
            this.array[this.top++] = x;
            return true;
        }
    }

    pop() {
        if (this.top < 0) {
            console.log("stackUnderflow");
            return 0;
        }
        else {
            let x = this.array[this.top--];
            return x;
        }
    }
}

function reverseString(str) {
    let n = str.length;
    let stack = new Stack(n);

    for (let i = 0; i < n; i++) {
        stack.push(str[i]);
    }

    for (let i = 0; i < n; i++) {
        stack = str[i].pop();
        str[i] = stack;
    }

}

let s = "Geeks";

s.split(" ");

reverseString(s);

console.log("reversed string: " + s.join(""));
